python/unblob/handlers/compression/geom_uzip.py

```diff
@@ -0,0 +1,79 @@
+from io import SEEK_SET
+from pathlib import Path
+from typing import Optional
+from zlib import decompress
+
+from structlog import get_logger
+
+from unblob.file_utils import Endian, FileSystem, StructParser, read_until_past
+from unblob.models import (
+    Extractor,
+    ExtractResult,
+    File,
+    HexString,
+    StructHandler,
+    ValidChunk,
+)
+
+""",
+The geom_uzip header follows the following structure:
+10 bytes shebang, with newline suffix: #!/bin/sh\n
+13 bytes version, with newline suffix: #V2.0 Format\n or #L3.0 Format\n
+105 bytes command, with null bytes suffix: (kldstat -qm g_uzip||kldload geom_uzip)>&-&&mount_cd9660 /dev/`mdconfig -af $0`.uzip $1\nexit $?\n\x00\x00\x00\x00\x00\x00\x00\x00\x00
+"""
+
+C_DEFINITIONS = r"""
+    typedef struct uzip_header{
+        char magic[10];             /* '10 bytes '*/
+        char version[13];           /* 13 bytes */
+        char format[105];           /* 105 bytes */
+        uint32_t block_size;        /* 4 bytes */
+        uint32_t block_count;       /* 4 bytes - Number of blocks */
+        uint64_t toc[block_count];  /* table of content */
+    } uzip_header_t;
+"""
+HEADER_STRUCT = "uzip_header_t"
+
+logger = get_logger()
+
+
+class UZIPExtractor(Extractor):
+    def extract(self, inpath: Path, outdir: Path):
+        infile = File.from_path(inpath)
+        parser = StructParser(C_DEFINITIONS)
+        header = parser.parse(HEADER_STRUCT, infile, Endian.BIG)
+        fs = FileSystem(outdir)
+        outpath = Path(inpath.stem)
+        with fs.open(outpath, "wb+") as outfile:
+            for current_offset, next_offset in zip(header.toc[:-1], header.toc[1:]):
+                compressed_len = next_offset - current_offset
+                if compressed_len == 0:
+                    continue
+                infile.seek(current_offset, SEEK_SET)
+                outfile.write(decompress(infile.read(compressed_len)))
+        return ExtractResult(reports=fs.problems)
+
+
+class UZIPHandler(StructHandler):
+    NAME = "uzip"
+    PATTERNS = [
+        HexString(
+            "23 21 2F 62 69 6E 2F 73 68 0A 23 (56 32 | 4c 33) 2e 30 20 46 6f 72 6d 61 74 0A"
+        )
+    ]
+
+    HEADER_STRUCT = HEADER_STRUCT
+    C_DEFINITIONS = C_DEFINITIONS
+    EXTRACTOR = UZIPExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, Endian.BIG)
+        # take the last TOC block offset, end of file is that block offset + block size,
+        # starting from the start offset
+        end_offset = start_offset + header.toc[-1]
+        file.seek(end_offset, SEEK_SET)
+        end_offset = read_until_past(file, b"\x00")
```

Comment:
can we just leave it there and let padding detection pick it up?

or are we sure that we can not read into something?

Reply:
for real world samples that we got here, it's usually standalone (no chunks that follows). This helps since extraction is done on the source file itself rather than carving then extraction.

Reply:
If it is standalone, why do we have potential padding?

Reply:
wrong choice of words on my end. in simpler terms: uzip format pads the compressed file with null byte so that they're aligned to a specific boundary. Padding is an inherent part of uzip. It's cleaner if we keep it rather than carve uzip and null padding separately.

Reply:
@rxpha3l can you try something like this ?

```
end_offset = round_up(start_offset + header.toc[-1], header.block_size)
```

and remove the lines 74 and 75.

Reply:
if you change the code you must re-generate your integration tests files otherwise they do not reflect the state of your code, and therefore break.

Reply:
after careful consideration and checking the different derivatives of this format, we decided to not try to include the padding suffix since it's not always there (not present with zstd).